### PR TITLE
feat: implement multi-org template gallery

### LINKED ISSUE #13775: Multi-org templates page
- [ ] Setup new /templates route to retrieve templates for all organizations for a user
- [ ] Create new template gallery page called `MultiOrgTemplatesPageView` using a grid layout
- [ ] Create Storybook component for `MultiOrgTemplatesPageView`
- [ ] Filter templates by organization-id
- [ ] Render multi-org template gallery page if multi-organization experiment is enabled
- [ ] Create component for `TemplateCard` in template gallery grid
- [ ] Create storybook component for `TemplateCard`
- [ ] Display organization count to the right of each organization in the organization labels list
- [ ] Highlight template card on mouseover, onClick links to template page
- [ ] Hide organization filter if only 1 organization exists

### PR DESCRIPTION
**Pull request by jaaydenh**
Resolves #13775 

- [x]  Setup new /templates route to retrieve templates for all organizations for a user
- [x]  Create new template gallery page called `MultiOrgTemplatesPageView` using a grid layout
- [x]  Create Storybook component for `MultiOrgTemplatesPageView`
- [x]  Filter templates by organization-id
- [x]  Render multi-org template gallery page if multi-organization experiment is enabled
- [x]  Create component for TemplateCard in template gallery grid
- [x]  Create storybook component for `TemplateCard`
- [x]  Display organization count to the right of each organization in the organization labels list
- [x]  Highlight template card on mouseover, onClick should link to template page
- [x]  Hide organization filter if the user has templates for only 1 organization

Screenshot 2024-07-16 at 8 11 56 PM

### CODE REVIEW

File: site/src/api/queries/templates.ts [hunk 1]
```diff
@@ -30,16 +31,26 @@ export const templateByName = (
...
+) => {
+  return {
+    queryKey: getTemplatesByOrganizationIdQueryKey(organizationId, deprecated),
+    queryFn: () =>
+      API.getTemplatesByOrganizationId(organizationId, { deprecated }),
+  };
+};
 
-export const templates = (organizationId: string, deprecated?: boolean) => {
+export const templates = (config: TemplateFilter = {}) => {
```
Comment (BrunoQuaresma) [inline site/src/api/queries/templates.ts:50]: Nit: Perhaps we should use 'filter' instead of 'config'?

File: site/src/api/typesGenerated.ts [hunk 2]
```diff
@@ -1202,7 +1202,7 @@ export interface TemplateExample {
 
 // From codersdk/organizations.go
 export interface TemplateFilter {
-  readonly OrganizationID: string;
+  readonly q?: string;
```
Comment (BrunoQuaresma) [inline site/src/api/typesGenerated.ts:1205]: Is this type missing the organization ID prop? I see it still there on https://github.com/coder/coder/pull/13784/files#diff-dbdae237ab2279354898d4a64bb02289c3b960ba644849f6fc96b15228b603d0R367.

**Comment by BrunoQuaresma** [comment]
I have a few design considerations. I believe we can avoid adding more components and create another view solely for filtering based on organizations. We could achieve this using pills as filters or by replicating what we have for workspaces.

Pills:

Pill Tabs

Replicating Workspaces:

Search

File: site/src/api/typesGenerated.ts (hunk 2 above)
Comment (jaaydenh) [reply site/src/api/typesGenerated.ts:1205]: In organizations.go, I believe typescript:"-" excludes OrganizationID from the generated types

**Review by aslilac — COMMENTED** [review]
sorry for taking so long to review! 💀

**Comment by jaaydenh** [comment]
> I have a few design considerations. I believe we can avoid adding more components and create another view solely for filtering based on organizations. We could achieve this using pills as filters or by replicating what we have for workspaces.
> 

@aslilac Can you comment on the feedback from @BrunoQuaresma as you have more context around the design decisions?

File: site/src/utils/templateAggregators.ts [hunk 3]
```diff
@@ -22,3 +23,20 @@ export const getTemplatesByTag = (
...
+    const org = template.organization_name;
+    if (orgs[org]) {
+      orgs[org].push(template);
+    } else {
+      orgs[org] = [template];
+    }
+  });
+
+  return orgs;
+};
```
[the remark this answers is not in the document]
Comment (jaaydenh) [reply site/src/utils/templateAggregators.ts]: @aslilac issue with "all" fixed in latest commit

**Comment by aslilac** [comment]
@BrunoQuaresma the design is based off of the starter template gallery, and aims to bring some distinction between the appearance of workspace related pages and the appearance of the template related pages. investigating the pills idea sounds interesting, but we can mess with the design more after the experiment has shipped.

**Review by aslilac — APPROVED** [review]

**Pull request merged**